core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js

```diff
@@ -0,0 +1,165 @@
+ */
+
+oppia.factory('TrainingDataEditorPanelService', [
+  '$rootScope', '$uibModal', 'AlertsService', 'UrlInterpolationService',
+  function($rootScope, $uibModal, AlertsService, UrlInterpolationService) {
+    return {
+      openTrainingDataEditor: function(externalSave) {
+        AlertsService.clearWarnings();
+        if (externalSave) {
+          $rootScope.$broadcast('externalSave');
+        }
+        $uibModal.open({
+          templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+            '/pages/exploration_editor/editor_tab/' +
+            'training_data_editor_directive.html'),
+          backdrop: true,
+          controller: [
+            '$scope', '$injector', '$uibModalInstance',
+            'ExplorationStatesService', 'EditorStateService',
+            'AnswerClassificationService', 'ExplorationContextService',
+            'stateInteractionIdService', 'AngularNameService',
+            'EXPLICIT_CLASSIFICATION', 'TRAINING_DATA_CLASSIFICATION',
+            'ExplorationHtmlFormatterService', 'ResponsesService',
+            'stateCustomizationArgsService', 'TrainingDataService',
+            'TrainingModalService',
+            function($scope, $injector, $uibModalInstance,
+                ExplorationStatesService, EditorStateService,
+                AnswerClassificationService, ExplorationContextService,
+                stateInteractionIdService, AngularNameService,
+                EXPLICIT_CLASSIFICATION, TRAINING_DATA_CLASSIFICATION,
+                ExplorationHtmlFormatterService, ResponsesService,
+                stateCustomizationArgsService, TrainingDataService,
+                TrainingModalService) {
+              var _explorationId = ExplorationContextService.getExplorationId();
+              var _stateName = EditorStateService.getActiveStateName();
+              var _state = ExplorationStatesService.getState(_stateName);
+              var answerGroupIndex = (
+                ResponsesService.getActiveAnswerGroupIndex());
+
+              $scope.stateContent = _state.content.getHtml();
+              $scope.inputTemplate = (
+                ExplorationHtmlFormatterService.getInteractionHtml(
+                  stateInteractionIdService.savedMemento,
+                  stateCustomizationArgsService.savedMemento,
+                  false, 'testInteractionInput'));
+              $scope.trainingData = [];
+
+              var _rebuildTrainingData = function() {
+                $scope.trainingData = [];
+                TrainingDataService.getTrainingDataOfAnswerGroup(
+                  answerGroupIndex).forEach(function(answer) {
+                  answerTemplate = (
+                    ExplorationHtmlFormatterService.getAnswerHtml(
+                      answer, stateInteractionIdService.savedMemento,
+                      stateCustomizationArgsService.savedMemento));
+                  $scope.trainingData.push({
+                    answer: answer,
```

Comment:
can we make a domain object for this?

Reply:
It is created only temporarily and also not used outside of this service. I don't think its worth creating a domain object since it is nothing concrete, just a temporarily created data structure so that view can be generated.

Reply:
Hm, technically, we're now planning to create domain objects for anything that's represented as a dict with fixed keys, just as a matter of principle. The problem with passing around raw dicts is that they tend to get modified and you then never really know what's in them at any given stage in time. (And it always starts off being a local small thing with limited use -- until it isn't.)

So -- still leaning towards a domain object, but I do take your point. Consider the above information as an additional data point, and feel free to make the call.

Reply:
I don't think that having a domain object is worth it since this structure is created only when the specific modal is opened (it's not even on the page). So I am skipping this.